src/ws/ws_manager.rs

```diff
@@ -94,27 +93,53 @@ impl WsManager {
     pub(crate) async fn new(url: String) -> Result<WsManager> {
         let stop_flag = Arc::new(AtomicBool::new(false));
 
-        let (ws_stream, _) = connect_async(url.clone())
-            .await
-            .map_err(|e| Error::Websocket(e.to_string()))?;
-
-        let (writer, mut reader) = ws_stream.split();
+        let (writer, mut reader) = Self::connect(url.as_str()).await?.split();
         let writer = Arc::new(Mutex::new(writer));
 
         let subscriptions_map: HashMap<String, Vec<SubscriptionData>> = HashMap::new();
         let subscriptions = Arc::new(Mutex::new(subscriptions_map));
         let subscriptions_copy = Arc::clone(&subscriptions);
 
         {
+            let writer = writer.clone();
             let stop_flag = Arc::clone(&stop_flag);
             let reader_fut = async move {
-                // TODO: reconnect
                 while !stop_flag.load(Ordering::Relaxed) {
                     let data = reader.next().await;
-                    if let Err(err) =
-                        WsManager::parse_and_send_data(data, &subscriptions_copy).await
-                    {
-                        error!("Error processing data received by WS manager reader: {err}");
+                    match data {
+                        Some(data) => {
+                            if let Err(err) = WsManager::parse_and_send_data(data, &subscriptions_copy).await {
+                                error!("Error processing data received by WsManager reader: {err}");
+                            }
+                        }
+                        None => {
+                            warn!("WsManager disconnected");
+                            if let Err(err) = WsManager::send_to_all_subscriptions(&subscriptions_copy, Message::NoData).await {
+                                warn!("Error sending disconnection notification err={err}");
+                            }
+                            tokio::time::sleep(Duration::from_secs(1)).await; //Always sleep for 1 second before attempting to reconnect so it does not spin reconnecting. This could be enhanced to exponential backoff.
```

Comment:
```suggestion
                            // Always sleep for 1 second before attempting to reconnect so it does not spin during reconnecting. This could be enhanced with exponential backoff.
                            tokio::time::sleep(Duration::from_secs(1)).await;
```